scripts/version_buckets.py

```diff
@@ -0,0 +1,236 @@
+
+This script should only be run once, when the following PR lands in master in
+the control panel:
+
+https://github.com/ministryofjustice/analytics-platform-control-panel/pull/816
+
+(From that moment on, all S3 buckets will be configured in this way by the
+control panel, as a default.)
+
+This script allows us to preview the potential changes and run the script on
+a defined number of buckets at once.
+"""
+import os
+import click
+import logging
+import sys
+import boto3
+import json
+from botocore.exceptions import ClientError
+from pathlib import Path
+
+
+# A flag to indicate if the command is to be run in verbose mode.
+VERBOSE = False
+
+
+# Setup logging. Logs will appear in user's home directory.
+LOGFILE = os.path.join(str(Path.home()), "bucket_update.log")
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logfile_handler = logging.FileHandler(LOGFILE)
+log_formatter = logging.Formatter("%(levelname)s: %(message)s")
+logfile_handler.setFormatter(log_formatter)
+logger.addHandler(logfile_handler)
+
+
+def check_bucket(name, bucket):
+    """
+    Returns a boolean indication to say if the referenced name/bucket is
+    associated with the control panel app.
+
+    They either start with the name "alpha-app-" or they have a tag of
+    "buckettype" with the value "datawarehouse".
+    See discussion here:
+
+    https://github.com/ministryofjustice/analytics-platform-ops/pull/320#discussion_r433765261
+    """
+    is_valid = name.startswith("alpha-app-")
+    if not is_valid:
+        try:
+            tags = bucket.Tagging()
+            tags.load()
+            for obj in tags.tag_set:
+                if obj["Key"] == "buckettype" and obj["Value"] == "datawarehouse":
+                    is_valid = True
+                    break
+        except:
+            is_valid = False
+    return is_valid
+
+
+def get_buckets(n=0):
+    """
+    Return a dictionary of {"bucket_name": bucket_object, } for the default
+    AWS account.
+    """
+    result = {}
+    click.echo("Getting S3 bucket details.")
+    s3 = boto3.client('s3')
+    bucket_metadata = s3.list_buckets()
+    if n:
+        buckets = bucket_metadata["Buckets"][:n]
+    else:
+        buckets = bucket_metadata["Buckets"]
+    with click.progressbar(buckets, len(buckets)) as bucket_list:
+        for bucket in bucket_list:
+            name = bucket["Name"]
+            bucket_obj = boto3.resource("s3").Bucket(name)
+            if check_bucket(name, bucket_obj):
+                versioning = bucket_obj.Versioning()
+                try:
+                    lifecycle_conf = bucket_obj.LifecycleConfiguration()
+                    lifecycle = lifecycle_conf.rules
+                except ClientError as ex:
+                    lifecycle = ""
+                result[name] = {
+                    "bucket": bucket_obj,
+                    "versioning": versioning.status,
+                    "lifecycle": lifecycle,
+                }
```

Comment:
No, because we only need to check tagging if the bucket doesn't meet the starts-with `alpha-app` requirement (i.e. saving on yet another network request).